### ISSUE TITLE
Some problems - tested with my laptop and musl libc

### ISSUE THREAD

**Issue by xhebox** [troglobit/finit#42]
**First, sorry for my poor english..**
And problems:

**1. reboot without sync**
in sig.c:
```
--- finit-3.0-rc1/sig.c 2016-09-14 04:49:41.000000000 +0800
+++ finit-3.0-rc1/sig.c 2016-09-14 04:49:41.000000000 +0800
@@ -152,6 +152,7 @@
        unmount_regular();

        /* We sit on / so we must remount it ro, try all the things! */
+       run("sync");
        run("/bin/mount -n -o remount,ro -t dummytype dummydev /");
        run("/bin/mount -n -o remount,ro dummydev /");
        run("/bin/mount -n -o remount,ro /");
```
I rebooted my laptop immediately after installing...crashed...I checked `/bin/finit`, found that it was empty.
Umount will sync automaticly but remount, so this was only happened on `/`.There is a need of sync command or soomething else can sync.

**2. didn't clean hidden files**
```
-       run_interactive("rm -rf /tmp/* /var/run/* /var/lock/*", "Cleaning up temporary directories");
+       run_interactive("rm -rf /tmp/* /tmp/.??* /var/run/* /var/lock/*", "Cleaning up temporary directories");
```
All files that named `.*` are not cleaned on boot.But `/tmp` has a lot of files named `.*`.

**3. start udevd but no trigger**
```
        run_interactive(devfsd, "Populating device tree");
+       run_interactive("/sbin/udevadm trigger --action=add --type=subsystems", "Trigger subsystems");
+       run_interactive("/sbin/udevadm trigger --action=add --type=devices", "Trigger devices");
```
With out trigger, kernel modules won't be loaded automaticly.Also devices is not under management.
For example, i915 was not loaded after boot, my network card was not renamed from `wlan0` to `wlp3s0` which means it was not handled by udev.

**4. missing dir**
```
+       makedir("/var/lock/subsys", 0755);
        makedir("/var/lock/subsys/messagebus", 0755);
```
No `subsys` in some distros.

**5. A question**
Can `rlimit` option be used on a specific process?This can be useful in cases like starting jack daemon.

:smiley: Thx for your excellent work on this init.

**Comment by troglobit** [comment]
This is brilliant, thank you so much!  :+1: 

I've committed and pushed the patches in your name, hope you don't mind.  Don't worry about your English, I'm not a native speaker either.  I welcome all contributions and usually the code speaks for itself :smiley: 

About `rlimit`.  At the moment it unfortunately it applies to *all* services, including Finit. It was mostly added to improve debugging of Finit and its plugins.  I'd like very much to add support for rlimit per process, as well as running processes in their own cgroups.  If you have any ideas for syntax and/or pull requests I'm all ears!

**Comment by troglobit** [comment]
@xhebox I hope this answered your question, please let me know if there is anything else I can do, or if we can close this issue.

**Comment by xhebox** [comment]
@troglobit Sorry for late, i'm still a student and it didn't matter what you did...Finit now works well.I'll open a new issue if i find something new.

**Comment by troglobit** [comment]
@xhebox No problem. Thanks for the feedback, and good luck! ☺🖖

**Comment by xhebox** [comment]
```
run_interactive("rm -rf /tmp/* /var/run/* /var/lock/*", "Cleaning up temporary directories");
run_interactive("rm -rf /tmp/* /tmp/.??* /var/run/* /var/lock/*", "Cleaning up temporary directories");
run("rm -rf /tmp/* /var/run/* /var/lock/*", "Cleaning up temporary directories");
run_("rm -rf /tmp/* /tmp/.??* /var/run/* /var/lock/*", "Cleaning up temporary directories");
```

These commands all failed in my tests...No files were removed.Patches did not solve the problem actually!But this worked(also what i did before open the issue):

```
system("rm -rf /tmp/* /tmp/.??* /var/run/* /var/lock/*", "Cleaning up temporary directories");
```

So, i thinks it should be `*`?I've noticed that `*` only appeared there.Then:

```
run_interactive("find /tmp /var/run /var/lock -mindepth 1 -exec rm -rf {} \;", "Cleaning up temporary directories");
```

Worked...See u tonight, time to school...

**Comment by troglobit** [comment]
On 2016-09-30 07:58, xhe wrote:
> |find /tmp /var/run /var/lock -mindepth 1 |

On my system I had to do add slashes ...

	find /tmp/ /var/run/ /var/lock/ -mindepth 1

... at least for /var/run and /var/lock

Good catch, again! :)
  /J

**Comment by xhebox** [comment]
So, just avoid using `*`? But i think using `find` is not good idea since there're a lot of diffrent implementations...Or, no `find` at all on embeded systems.

**Comment by troglobit** [comment]
Ya, I'm looking now at a replacement using recursive `glob(3)` written in C.

**Comment by troglobit** [comment]
Here's a proposal, I'll test it out and commit during the weekend :+1: 

```C
diff --git a/finit.c b/finit.c
index e1388a9..0727259 100644
--- a/finit.c
+++ b/finit.c
@@ -25,6 +25,7 @@
 #include "config.h"            /* Generated by configure script */
 
 #include <ctype.h>
+#include <ftw.h>
 #include <glob.h>
 #include <dirent.h>
 #ifdef HAVE_FSTAB_H
@@ -117,6 +118,27 @@ static int fsck(int pass)
        return 0;
 }
 
+static int show(const char *fpath, const struct stat *UNUSED(sb), int UNUSED(tflag), struct FTW *UNUSED(ftwbuf))
+{
+       _d("Removing %s", fpath);
+       remove(fpath);
+
+       return 0;
+}
+
+static void bootclean(void)
+{
+       char *dir[] = {
+               "/tmp/",
+               "/var/run/",
+               "/var/lock/",
+               NULL
+       };
+
+       for (int i = 0; dir[i]; i++)
+               nftw(dir[i], show, 20, FTW_DEPTH);
+}
+
 static void networking(void)
 {
        size_t i;
@@ -378,7 +400,7 @@ int main(int argc, char* argv[])
        umask(0022);
 
        /* Cleanup of stale files, if any still linger on. */
-       run_interactive("rm -rf /tmp/* /tmp/.??* /var/run/* /var/lock/*", "Cleaning up temporary directories");
+       bootclean();
 
        /* Base FS up, enable standard SysV init signals */
        sig_setup(&loop);
```

**Comment by troglobit** [comment]
Should work now, please reopen if the problem persists.
Thank you for contributing to Open Source! <3

**Comment by xhebox** [comment]
Worked..But remove the dir itself..
The output of `nftw`, as you can see, `/tmp`...So `remove("/tmp");` is called...
```
/tmp/serverauth.J28auBywHb
/tmp/.X0-lock
/tmp/.org.chromium.Chromium.cpDcEc/SingletonSocket
/tmp/.org.chromium.Chromium.cpDcEc/SingletonCookie
/tmp/.org.chromium.Chromium.cpDcEc
/tmp/.X11-unix/X0
/tmp/.X11-unix
/tmp/fcitx-socket-:0
/tmp/pulse-aqZdS4kW8qxn/pid
/tmp/pulse-aqZdS4kW8qxn/native
/tmp/pulse-aqZdS4kW8qxn
/tmp/
```

**Comment by troglobit** [comment]
Woups! ;-)

(Fix coming soon :)

**Comment by xhebox** [comment]
fresh patch..
```
--- finit-master/finit.c        2016-10-02 11:59:16.980643479 +0800
+++ finit-master/finit.c        2016-10-02 11:59:16.980643479 +0800
@@ -118,8 +118,10 @@
        return 0;
 }

-static int do_clean(const char *fpath, const struct stat *UNUSED(sb), int UNUSED(tflag), struct FTW *UNUSED(ftwbuf))
+static int do_clean(const char *fpath, const struct stat *UNUSED(sb), int UNUSED(tflag), struct FTW *(ftwbuf))
 {
+       if(ftwbuf->level == 0)
+               return 0;
        _d("Removing %s", fpath);
        remove(fpath);
```

PS: Urandom gave me a `[FAIL]`.I guess it's a `missing dir`.Testing.

**Comment by troglobit** [comment]
Heh, have the exact same patch here. Didn't push it yesterday evening (CET)
though :-)

**Comment by xhebox** [comment]
And your libite's code needs to be fixed :D.

2016-10-02 16:55 GMT+08:00 Joachim Nilsson <notifications@github.com>:

> Heh, have the exact same patch here. Didn't push it yesterday evening (CET)
> though :-)
>

**Comment by troglobit** [comment]
How come, what did you find? :)

**Comment by troglobit** [comment]
Ah, now I see!  

Yeah, a new issue would be great, can you add one?

I'll push the ftw() fix to this one, so it can be closed.

**Comment by troglobit** [comment]
There, pushed fix and also opened issue #46 about the permissions problem.
